### ISSUE TITLE
Gate.io watch_orderbook error "orderbook update has a nonce bigger than u"

### ISSUE THREAD

**Issue by Tongoko** [ccxt/ccxt#17733]
### Operating System

Windows

### Programming Languages

python

### CCXT Version

3.0.80

### Description

I have recently upgraded ccxt from 2.0.68 to 3.0.80.
Before the update, watch_orderbook was working flawlessly. After the update, there is an error coming up regularly saying "orderbook update has a nonce bigger than u"

I let it run for a couple hours and noticed that the error has a tendency to show up more often on full hours like 3:00:00, 15:00:00 etc (about 1x / 3 hours on average)
And additionaly, it's also happening in random intervals in between.

I tried a couple workarounds, but the problem still persists:

> Try, except & continue logic works sometimes, but quite often it causes it to infinitely try unable to retrieve the orderbook.
> I added a snapshotDelay: 20, which doesn't change the frequency of the error much.

It checked watchOrderbook code for both versions and it seems there have been quite some changes in the code between the two. Maybe something has been broken?

### Code

```
await ccxtpro.gate().watchOrderBook()

```

**Comment by carlosmiei** [comment]
Hello @Tongoko, thanks for reporting it, we're currently investigating this issue and the provided information might be very useful. We will let you know as soon as we get an update.

**Comment by Tongoko** [comment]
@carlosmiei Cool, awesome!

**Comment by pcriadoperez** [comment]
Hi @Tongoko , I've been trying to reproduce the error but haven't been susccesful.

I looked at the previous implementation and looks like the current implementation is more strict checking the nonce, which can explain why you see the orderbook fall out of sync sometimes at peak hours and why it didn't happen in the past.

In any case it should be able to handle the error and reconnect. So I thought there could be an issue recovering from the error so I wrote a script that would throw the nonce error after 1 minute.  However on my side it is able to reconnect fine with for both spot and swap symbols with the default snapshot delay.

I assume you are calling it inside a while (true), try, catch clause as the examples? Is there any other details you can think of to try and reproduce from my side?

**Comment by Tongoko** [comment]
@pcriadoperez That's correct, while True catch errors. Have you tried running watch_orderbook for a couple hours?

I've been able to get it to reconnect consistently too by pausing the code for a second when catching the error, before syncing order_book again.

I've been running watch_orderbook the last couple days and the nonce bigger than u error seems to have stopped happening in between the full hours.

However the error is still happening on every couple full hour. Like 03:00 05:00 06:00 08:00...

Do you think that is also related to the load happening on the full hours on the exchange side?

**Comment by Tongoko** [comment]
![16835235197166088534370183576329](https://user-images.githubusercontent.com/117158948/236741007-bd926714-cc56-4d7c-885d-a10e47d491d8.jpg)

Here is an example from last nightbpicked up by my log

**Comment by pcriadoperez** [comment]
Yes, it could be something from the exchange. And being so consistent at
those times it makes me think it's something at their end. The way to
validate it, would be to turn on verbose output and check the raw messages
around the time of the error. I'll try again today to keep a log and see if
I catch it during those errors.

**Comment by spc789** [comment]
@pcriadoperez sometimes it happened on binance too (as described in one of my closed issues). 
Is there a way to get the verbose output to a file instead of to the screen ?
If so, I'm happy to let the output run for a couple of hours even days so that the possible issue can be debugged more quickly ?

**Comment by kroitor** [comment]
Hello @Tongoko, closing this during a cleanup — the failure mode you hit was redesigned away: the gate orderbook stream now runs a proper nonce state machine, stale deltas are dropped, contiguous deltas are applied, and a sequence gap tears down the subscription and the book and resyncs from a fresh snapshot, optionally surfacing a structured ChecksumError instead of the raw nonce error you saw, with a snapshotDelay option controlling the initial snapshot timing. If a book still desyncs on a current version, a fresh issue with a verbose log would be very welcome. Thanks for the report back then!
